lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,331 @@ using namespace mlir;
+  return std::make_pair(indices, reverseIndices);
+}
+
+auto getLoopVarIndicesToKeep(scf::ForOp loop, const Partition *partition,
+                             const WarpSchedule &schedule) {
+  auto loopVarCategories = classifyLoopVars(loop, partition, schedule);
+  return getLoopVarIndicesToKeep(loop, partition, loopVarCategories);
+}
+
+const Partition *getPartition(Operation *op, const WarpSchedule &schedule) {
+  while (op && !schedule.getPartition(op)) {
+    op = op->getParentOp();
+  }
+  if (op) {
+    return schedule.getPartition(op);
   }
-  eraseLoopCarriedValues(loop, std::move(toErase));
-  // Erase the schedule attributes.
-  WarpSchedule::eraseFrom(loop);
+  return nullptr;
 }
 
-// Given a loop and a scheduled partition, slice a copy of the partition into a
-// new loop. This returns the block containing the new loop.
-static FailureOr<std::unique_ptr<Block>>
-slicePartition(scf::ForOp baseLoop, const WarpSchedule &baseSchedule,
-               const Partition *slicePartition) {
-  // Generate the partition loop by cloning the whole loop and deleting anything
-  // that doesn't belong to the partition and the root partition. This is easier
-  // than trying to generate a new loop from scratch while keeping the
-  // operations in the same order.
-  scf::ForOp loop = baseLoop.clone();
-  std::unique_ptr<Block> block = std::make_unique<Block>();
-  block->push_back(loop);
-  WarpSchedule schedule = *WarpSchedule::deserialize(loop);
-  Partition *partition = schedule.getPartition(slicePartition->getIndex());
-
-  // Check for results that need to be passed back into the default warp group.
-  SmallVector<unsigned> resultIndices;
-  baseSchedule.iterateOutputs(
-      baseLoop, slicePartition, [&](Operation *op, OpOperand &use) {
-        unsigned idx = use.getOperandNumber();
-        // Ignore results with no uses.
-        if (isa<scf::YieldOp>(op) && !baseLoop.getResult(idx).use_empty())
-          resultIndices.push_back(idx);
-      });
-
-  // Pass these results through shared memory.
-  for (unsigned resultIdx : resultIndices) {
-    Value result = loop.getResult(resultIdx);
-    auto ty = dyn_cast<RankedTensorType>(result.getType());
-    if (!ty) {
-      return mlir::emitWarning(
-          result.getLoc(),
-          "FIXME: only tensor partition results are supported");
+void cloneOpsInBlock(Block *block, SmallVector<WarpGroupBuilder> &builders,
+                     const WarpSchedule &schedule);
+
+void cloneForOp(scf::ForOp forOp, SmallVector<WarpGroupBuilder> &builders,
+                const WarpSchedule &schedule) {
+  SmallVector<scf::ForOp> newForOps;
+  for (auto [b, partition] : llvm::zip(builders, schedule.getPartitions())) {
+    auto [newLoopIndices, _] =
+        getLoopVarIndicesToKeep(forOp, &partition, schedule);
+    auto lb = b.mapping.lookupOrDefault(forOp.getLowerBound());
+    auto ub = b.mapping.lookupOrDefault(forOp.getUpperBound());
+    auto step = b.mapping.lookupOrDefault(forOp.getStep());
```

Comment:
What happens when the lookup returns the default `Value()`? If you expect the lookup to succeed, it would be better to use `lookup`

Reply:
You're right. I misunderstood what `lookupOrDefault` does.